[R2] Add extensions and unicode interop documentation (#4330)
cloudflare/cloudflare-docs: content/r2/platform/s3-compatibility/extensions.md

---
title: Extensions
pcx-content-type: reference
meta:
  title: Extensions
---

# Extensions

R2 implements some extensions on top of the stock S3 API. This page outlines those additional features that are available:

## Extended Metadata Using Unicode

The [Cloudflare Workers R2 API](/workers/runtime-apis/r2) supports Unicode in keys and values natively without any additional
encoding/decoding needed for the `customMetadata` field. These fields map 1:1 to the headers that are prefixed with `x-amz-meta-`

> in the R2 S3-compatible API endpoint.

Reviewer: We dont have this Runtime API page btw, so linking to it breaks CI

> in the R2 S3-compatible API endpoint.

Author: Yup - I'll rebase once @greg-mckeon pushes the right page

> To bridge the gap that HTTP headers can only have ASCII keys and values, all `x-amz-meta-` prefixed header values are RFC2047 decoded
> before being stored in R2. Values with unicode are RFC2047 encoded before rendering back responses. The length limit on metadata
> values applies on the decoded Unicode value.

Reviewer: ```suggestion
HTTP header names and values may only contain ASCII characters, which is a small subset of the Unicode character library. To easily accommodate users, R2 adheres to [RFC2047](https://datatracker.ietf.org/doc/html/rfc2047) and automatically decodes all `x-amz-meta-*` header values before storage. On retrieval, any metadata values with unicode are RFC2047-encoded before rendering the response. The length limit for metadata values is applied to the decoded Unicode value.
```

> Be mindful of this when using both Workers and S3 API endpoints to access the same data. If the R2 metadata keys contain Unicode then
> they are stripped when returning responses through S3 and the `x-amz-missing-meta` header is set to the number of keys that were omitted.

Reviewer: ```suggestion
{{<Aside type="warning" header="Metadata variance">
Be mindful when using both Workers and S3 API endpoints to access the same data. 

If the R2 metadata keys contain Unicode, they are stripped when accessed through the through S3 API and the `x-amz-missing-meta` header is set to the number of keys that were omitted.
{{</Aside>}}
```

> Unicode is not allowed in normal HTTP headers and those headers must always follow standard rules for HTTP. These headers map to the
> `httpMetadata` field in the R2 bindings:

Reviewer: Stated above, dont think need to repeat

```suggestion
These headers map to the `httpMetadata` field in the R2 bindings:
```

* `Content-Encoding`
* `Content-Type`
* `Content-Language`
* `Content-Disposition`
* `Cache-Control`

> * `Expires`

Reviewer: I think this would be better served as a table with the `HTTP header` and `property name` columns

If using Unicode in object key names, it may be helpful to review the [technical notes](/r2/platform/unicode-interoperability) regarding
that.

## CopyObject

### MERGE metadata directive

> The `x-amz-metadata-directive` can be given a value of `MERGE` (in addition to the standard `COPY`/`REPLACE` versions). This is
> a combination of `COPY` + `REPLACE` which will `COPY` any metadata keys from the source object and `REPLACE` any that are specified
> in the request with the new value. This cannot be used currently to merge in deletion of existing metadata keys. For that you still
> need to use `REPLACE`.

Reviewer: ```suggestion
The `x-amz-metadata-directive` allows a `MERGE` value, in addition to the standard `COPY` and `REPLACE` options. When used, `MERGE` is a combination of `COPY` and `REPLACE`, which will `COPY` any metadata keys from the source object and `REPLACE` those that are specified in the request with the new value. You cannot use `MERGE` to remove existing metadata keys from the source — use `REPLACE` instead.
```

---
title: Extensions
pcx-content-type: reference
meta:
  title: Extensions
---

# Extensions

R2 implements some extensions on top of the stock S3 API. This page outlines those additional features that are available:

## Extended Metadata Using Unicode

> The [Workers R2 API](/r2/api) supports Unicode in supports Unicode in keys and values natively without requiring any additional
> encoding or decoding for the `customMetadata` field. These fields map to the `x-amz-meta-`-prefixed headers used within the
> R2 S3-compatible API endpoint.

Reviewer: ```suggestion
The [Workers R2 API](/r2/api/) supports Unicode in keys and values natively without requiring any additional encoding or decoding for the `customMetadata` field. These fields map to the `x-amz-meta-`-prefixed headers used within the R2 S3-compatible API endpoint.
```

HTTP header names and values may only contain ASCII characters, which is a small subset of the Unicode character library. To easily
accommodate users, R2 adheres to [RFC2047](https://datatracker.ietf.org/doc/html/rfc2047) and automatically decodes all

> `x-amz-meta-*` header values before storage. On retrieval, any metadata values with unicode are RFC2047-encoded before rendering
> the response. The length limit for metadata values is applied to the decoded Unicode value.

Reviewer: ```suggestion
`x-amz-meta-*` header values before storage. On retrieval, any metadata values with unicode are RFC2047-encoded before rendering the response. The length limit for metadata values is applied to the decoded Unicode value.
```

{{<Aside type="warning" header="Metadata variance">

> Be mindful when using both Workers and S3 API endpoints to access the same data. If the R2 metadata keys contain Unicode, they
> are stripped when accessed through the through S3 API and the `x-amz-missing-meta` header is set to the number of keys that were
> omitted.

Reviewer: ```suggestion
Be mindful when using both Workers and S3 API endpoints to access the same data. If the R2 metadata keys contain Unicode, they are stripped when accessed through the through S3 API and the `x-amz-missing-meta` header is set to the number of keys that were omitted.
```

{{</Aside>}}

These headers map to the `httpMetadata` field in the R2 bindings:

> {{<table-wrap>}}

Reviewer: ```suggestion

{{<table-wrap>}}
```

| HTTP Header           | Property Name                     |
|---------------------- | --------------------------------- |
| `Content-Encoding`    | `httpMetadata.contentEncoding`    |
| `Content-Type`        | `httpMetadata.contentType`        |
| `Content-Language`    | `httpMetadata.contentLanguage`    |
| `Content-Disposition` | `httpMetadata.contentDisposition` |
| `Cache-Control`       | `httpMetadata.cacheControl`       |
| `Expires`             | `httpMetadata.expires`            |
{{</table-wrap>}}

If using Unicode in object key names, it may be helpful to review the [technical notes](/r2/platform/unicode-interoperability) regarding
that.

## CopyObject

### MERGE metadata directive

> The `x-amz-metadata-directive` allows a `MERGE` value, in addition to the standard `COPY` and `REPLACE` options. When used,
> `MERGE` is a combination of `COPY` and `REPLACE`, which will `COPY` any metadata keys from the source object and `REPLACE` those
> that are specified in the request with the new value. You cannot use `MERGE` to remove existing metadata keys from the
> source — use `REPLACE` instead.

Reviewer: ```suggestion
The `x-amz-metadata-directive` allows a `MERGE` value, in addition to the standard `COPY` and `REPLACE` options. When used, `MERGE` is a combination of `COPY` and `REPLACE`, which will `COPY` any metadata keys from the source object and `REPLACE` those that are specified in the request with the new value. You cannot use `MERGE` to remove existing metadata keys from the source — use `REPLACE` instead.
```
